File: modules/ark-ocr/android/src/main/java/expo/modules/arkocr/ArkOcrModule.kt
package expo.modules.arkocr

import android.graphics.Bitmap
import android.graphics.Color
import android.graphics.pdf.PdfRenderer
import android.net.Uri
import android.os.ParcelFileDescriptor
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.text.TextRecognition
import com.google.mlkit.vision.text.latin.TextRecognizerOptions
import com.tom_roush.pdfbox.android.PDFBoxResourceLoader
import com.tom_roush.pdfbox.pdmodel.PDDocument
import com.tom_roush.pdfbox.text.PDFTextStripper
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.io.File
import kotlin.math.max
import kotlin.math.min

class ArkOcrModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("ArkOcr")

    AsyncFunction("recognizeText") { uri: String, promise: Promise ->
      val context = appContext.reactContext
        ?: return@AsyncFunction promise.reject(
          "ERR_CONTEXT_UNAVAILABLE",
          "Android context is not available for OCR.",
          null
        )

      val imageUri = if (uri.startsWith("file://") || uri.startsWith("content://")) {
        Uri.parse(uri)
      } else {
        Uri.fromFile(File(uri))
      }

      val image = try {
        InputImage.fromFilePath(context, imageUri)
      } catch (error: Exception) {
        promise.reject("ERR_IMAGE_UNREADABLE", "Could not read image for OCR.", error)
        return@AsyncFunction
      }

      val recognizer = TextRecognition.getClient(TextRecognizerOptions.DEFAULT_OPTIONS)
      recognizer.process(image)
        .addOnSuccessListener { visionText ->
          val blocks = visionText.textBlocks.map { block ->
            mapOf(
              "text" to block.text,
              "confidence" to block.lines
                .flatMap { line -> line.elements }
                .mapNotNull { element -> element.confidence.takeIf { it >= 0f } }
                .average()
                .takeIf { !it.isNaN() }
            )
          }
          promise.resolve(
            mapOf(
              "text" to visionText.text,
              "blocks" to blocks
            )
          )
        }
        .addOnFailureListener { error ->
          promise.reject("ERR_OCR_FAILED", "Text recognition failed.", error)
        }
    }

    AsyncFunction("extractPdfText") { uri: String, maxPages: Int, promise: Promise ->
      val context = appContext.reactContext
        ?: return@AsyncFunction promise.reject(
          "ERR_CONTEXT_UNAVAILABLE",
          "Android context is not available for PDF extraction.",
          null
        )
      val file = fileFromUri(uri)
      if (!file.exists()) {
        promise.reject("ERR_PDF_NOT_FOUND", "PDF file is not available.", null)
        return@AsyncFunction
      }

      try {
        PDFBoxResourceLoader.init(context)
        PDDocument.load(file).use { document ->
          val pageCount = document.numberOfPages
          val pageLimit = min(max(maxPages, 1), pageCount)
          val stripper = PDFTextStripper()
          val pages = mutableListOf<Map<String, Any?>>()
          for (pageNumber in 1..pageLimit) {
            stripper.startPage = pageNumber
            stripper.endPage = pageNumber
            pages.add(
              mapOf(
                "pageNumber" to pageNumber,
                "text" to stripper.getText(document).trim(),
                "extractionMethod" to "text_layer",
                "confidence" to null
              )
            )
          }
          promise.resolve(
            mapOf(
              "pageCount" to pageCount,
              "pages" to pages,
              "truncated" to (pageLimit < pageCount)
            )
          )
        }
      } catch (error: Exception) {
        promise.reject("ERR_PDF_TEXT_FAILED", "PDF text extraction failed.", error)
      }
    }

    AsyncFunction("recognizePdf") { uri: String, maxPages: Int, renderDpi: Int, promise: Promise ->
      val context = appContext.reactContext
        ?: return@AsyncFunction promise.reject(
          "ERR_CONTEXT_UNAVAILABLE",
          "Android context is not available for PDF OCR.",
          null
        )
      val file = fileFromUri(uri)
      if (!file.exists()) {
        promise.reject("ERR_PDF_NOT_FOUND", "PDF file is not available.", null)
        return@AsyncFunction
      }

      val descriptor = try {
        ParcelFileDescriptor.open(file, ParcelFileDescriptor.MODE_READ_ONLY)
      } catch (error: Exception) {
        promise.reject("ERR_PDF_UNREADABLE", "Could not open PDF for OCR.", error)
        return@AsyncFunction
      }
      val renderer = try {
        PdfRenderer(descriptor)
      } catch (error: Exception) {
        descriptor.close()
        promise.reject("ERR_PDF_RENDERER", "Could not render PDF pages for OCR.", error)
        return@AsyncFunction
      }

      val recognizer = TextRecognition.getClient(TextRecognizerOptions.DEFAULT_OPTIONS)
      val totalPages = renderer.pageCount
      val pageLimit = min(max(maxPages, 1), totalPages)
      val pages = mutableListOf<Map<String, Any?>>()

      fun closeRenderer() {
        try {
          renderer.close()
        } catch (_: Exception) {
        }
        try {
          descriptor.close()
        } catch (_: Exception) {
        }
      }

      fun processPage(index: Int) {
        if (index >= pageLimit) {
          closeRenderer()
          promise.resolve(
            mapOf(
              "pageCount" to totalPages,
              "pages" to pages,
              "truncated" to (pageLimit < totalPages)
            )
          )
          return
        }

        val page = renderer.openPage(index)
        val scale = max(renderDpi, 96).toFloat() / 72f
        val width = max((page.width * scale).toInt(), 1)
        val height = max((page.height * scale).toInt(), 1)
        val bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
        bitmap.eraseColor(Color.WHITE)
        page.render(bitmap, null, null, PdfRenderer.Page.RENDER_MODE_FOR_DISPLAY)
        page.close()

        recognizer.process(InputImage.fromBitmap(bitmap, 0))
          .addOnSuccessListener { visionText ->
            pages.add(
              mapOf(
                "pageNumber" to (index + 1),
                "text" to visionText.text.trim(),
                "extractionMethod" to "ocr",
                "confidence" to null
              )
            )
            bitmap.recycle()
            processPage(index + 1)
          }
          .addOnFailureListener { error ->
            bitmap.recycle()
            closeRenderer()
            promise.reject("ERR_PDF_OCR_FAILED", "PDF OCR failed.", error)
          }
      }

      processPage(0)
    }
  }

  private fun fileFromUri(uri: String): File {
    return if (uri.startsWith("file://")) {
      File(Uri.parse(uri).path ?: uri)
    } else {
      File(uri)
    }
  }
}

File: modules/ark-ocr/ios/ArkOcrModule.swift
import CoreGraphics
import ExpoModulesCore
import PDFKit
import UIKit
import Vision

public class ArkOcrModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ArkOcr")

    AsyncFunction("recognizeText") { (uri: String, promise: Promise) in
      DispatchQueue.global(qos: .userInitiated).async {
        do {
          let image = try Self.loadImage(from: uri)
          let result = try Self.recognize(cgImage: image)
          promise.resolve([
            "text": result.text,
            "blocks": result.blocks
          ])
        } catch {
          promise.reject(Self.exception(from: error, fallbackCode: "ERR_OCR_FAILED", fallbackReason: "Text recognition failed."))
        }
      }
    }

    AsyncFunction("extractPdfText") { (uri: String, maxPages: Int, promise: Promise) in
      DispatchQueue.global(qos: .userInitiated).async {
        do {
          let document = try Self.loadPdf(from: uri)
          let pageCount = document.pageCount
          let pageLimit = min(max(maxPages, 1), pageCount)
          var pages: [[String: Any?]] = []

          if pageLimit > 0 {
            for pageIndex in 0..<pageLimit {
              let page = document.page(at: pageIndex)
              pages.append([
                "pageNumber": pageIndex + 1,
                "text": page?.string?.trimmingCharacters(in: .whitespacesAndNewlines) ?? "",
                "extractionMethod": "text_layer",
                "confidence": nil
              ])
            }
          }

          promise.resolve([
            "pageCount": pageCount,
            "pages": pages,
            "truncated": pageLimit < pageCount
          ])
        } catch {
          promise.reject(Self.exception(from: error, fallbackCode: "ERR_PDF_TEXT_FAILED", fallbackReason: "PDF text extraction failed."))
        }
      }
    }

    AsyncFunction("recognizePdf") { (uri: String, maxPages: Int, renderDpi: Int, promise: Promise) in
      DispatchQueue.global(qos: .userInitiated).async {
        do {
          let document = try Self.loadPdf(from: uri)
          let pageCount = document.pageCount
          let pageLimit = min(max(maxPages, 1), pageCount)
          var pages: [[String: Any?]] = []

          if pageLimit > 0 {
            for pageIndex in 0..<pageLimit {
              guard let page = document.page(at: pageIndex) else {
                continue
              }
              let image = try Self.render(page: page, dpi: renderDpi)
              let result = try Self.recognize(cgImage: image)
              pages.append([
                "pageNumber": pageIndex + 1,
                "text": result.text.trimmingCharacters(in: .whitespacesAndNewlines),
                "extractionMethod": "ocr",
                "confidence": nil
              ])
            }
          }

          promise.resolve([
            "pageCount": pageCount,
            "pages": pages,
            "truncated": pageLimit < pageCount
          ])
        } catch {
          promise.reject(Self.exception(from: error, fallbackCode: "ERR_PDF_OCR_FAILED", fallbackReason: "PDF OCR failed."))
        }
      }
    }
  }

  private static func loadImage(from uri: String) throws -> CGImage {
    let url = try fileURL(from: uri)
    guard let image = UIImage(contentsOfFile: url.path)?.cgImage else {
    throw Exception(name: "ERR_IMAGE_UNREADABLE", description: "Could not read image for OCR.")
    }
    return image
  }

  private static func loadPdf(from uri: String) throws -> PDFDocument {
    let url = try fileURL(from: uri)
    guard FileManager.default.fileExists(atPath: url.path) else {
      throw Exception(name: "ERR_PDF_NOT_FOUND", description: "PDF file is not available.")
    }
    guard let document = PDFDocument(url: url) else {
      throw Exception(name: "ERR_PDF_UNREADABLE", description: "Could not open PDF.")
    }
    return document
  }

  private static func fileURL(from uri: String) throws -> URL {
    if uri.hasPrefix("file://"), let url = URL(string: uri) {
      return url
    }
    if uri.hasPrefix("/") {
      return URL(fileURLWithPath: uri)
    }
    throw Exception(name: "ERR_FILE_URI_UNSUPPORTED", description: "Only local file URIs are supported for OCR.")
  }

  private static func render(page: PDFPage, dpi: Int) throws -> CGImage {
    let bounds = page.bounds(for: .mediaBox)
    let scale = max(CGFloat(dpi), 96) / 72
    let size = CGSize(width: max(bounds.width * scale, 1), height: max(bounds.height * scale, 1))
    let format = UIGraphicsImageRendererFormat.default()
    format.scale = 1
    let renderer = UIGraphicsImageRenderer(size: size, format: format)
    let image = renderer.image { context in
      UIColor.white.set()
      context.fill(CGRect(origin: .zero, size: size))
      context.cgContext.translateBy(x: 0, y: size.height)
      context.cgContext.scaleBy(x: scale, y: -scale)
      context.cgContext.translateBy(x: -bounds.origin.x, y: -bounds.origin.y)
      page.draw(with: .mediaBox, to: context.cgContext)
    }
    guard let cgImage = image.cgImage else {
      throw Exception(name: "ERR_PDF_RENDERER", description: "Could not render PDF pages for OCR.")
    }
    return cgImage
  }

  private static func recognize(cgImage: CGImage) throws -> (text: String, blocks: [[String: Any?]]) {
    var recognizedText = ""
    var recognizedBlocks: [[String: Any?]] = []
    var requestError: Error?
    let semaphore = DispatchSemaphore(value: 0)

    let request = VNRecognizeTextRequest { request, error in
      defer { semaphore.signal() }
      if let error {
        requestError = error
        return
      }
      let observations = (request.results as? [VNRecognizedTextObservation]) ?? []
      let texts = observations.compactMap { observation -> String? in
        observation.topCandidates(1).first?.string.trimmingCharacters(in: .whitespacesAndNewlines)
      }.filter { !$0.isEmpty }

      recognizedText = texts.joined(separator: "\n")
      recognizedBlocks = observations.compactMap { observation in
        guard let candidate = observation.topCandidates(1).first else {
          return nil
        }
        let text = candidate.string.trimmingCharacters(in: .whitespacesAndNewlines)
        if text.isEmpty {
          return nil
        }
        return [
          "text": text,
          "confidence": Double(candidate.confidence)
        ]
      }
    }
    request.recognitionLevel = .accurate
    request.usesLanguageCorrection = true

    let handler = VNImageRequestHandler(cgImage: cgImage, options: [:])
    try handler.perform([request])
    semaphore.wait()

    if let requestError {
      throw requestError
    }
    return (recognizedText, recognizedBlocks)
  }

  private static func exception(from error: Error, fallbackCode: String, fallbackReason: String) -> Exception {
    if let exception = error as? Exception {
      return exception
    }
    return Exception(name: fallbackCode, description: error.localizedDescription.isEmpty ? fallbackReason : error.localizedDescription)
  }
}
